File: src-tauri/src/whisper.rs
use reqwest::{multipart, Client};
use serde::Deserialize;
use std::time::Duration;
use regex::Regex;

#[derive(Debug, Deserialize)]
pub struct WhisperResponse {
    pub text: String,
}

#[derive(Clone)]
pub struct WhisperClient {
    client: Client,
    url: String,
}

impl WhisperClient {
    pub fn new(url: String) -> Result<Self, String> {
        let client = Client::builder()
            .timeout(Duration::from_secs(60))
            .build()
            .map_err(|e| format!("Failed to create HTTP client: {}", e))?;

        Ok(Self { client, url })
    }

    pub async fn transcribe(&self, audio_data: &[u8], remove_filler_words: bool) -> Result<String, String> {
        let part = multipart::Part::bytes(audio_data.to_vec())
            .file_name("audio.wav")
            .mime_str("audio/wav")
            .map_err(|e| format!("Failed to create mime part: {}", e))?;

        let mut form = multipart::Form::new()
            .part("file", part)
            .text("temperature", "0.2")
            .text("response_format", "json");

        if remove_filler_words {
            form = form.text("prompt", "The following is a clear, formal transcript without any stutters, repetitions, or filler words like um and ah.");
        }

        let response = self
            .client
            .post(&self.url)
            .multipart(form)
            .send()
            .await
            .map_err(|e| format!("Failed to send request: {}", e))?;

        if !response.status().is_success() {
            let status = response.status();
            let error_text = response
                .text()
                .await
                .unwrap_or_else(|_| "Unknown error".to_string());
            return Err(format!("Whisper API error: {} - {}", status, error_text));
        }

        let whisper_response: WhisperResponse = response
            .json()
            .await
            .map_err(|e| format!("Failed to parse response: {}", e))?;

        Ok(whisper_response.text.trim().to_string())
    }

    pub fn set_url(&mut self, url: String) {
        self.url = url;
    }
}

pub fn clean_filler_words(text: &str) -> String {
    let filler_words_pattern = Regex::new(r"(?i)\b(um|uh|ah|er|hmm|mm)\b\.?").unwrap();
    let mut cleaned = filler_words_pattern.replace_all(text, "").to_string();
    
    let double_space = Regex::new(r"\s+").unwrap();
    cleaned = double_space.replace_all(&cleaned, " ").to_string();
    
    let leading_trailing_space = Regex::new(r"^\s+|\s+$").unwrap();
    cleaned = leading_trailing_space.replace_all(&cleaned, "").to_string();
    
    let punctuation_fix = Regex::new(r"\s+([.,!?;])").unwrap();
    cleaned = punctuation_fix.replace_all(&cleaned, "$1").to_string();
    
    cleaned
}

File: src/main/whisper.ts
import type { AppConfig } from '../types/ipc';

export async function transcribe(audioData: Uint8Array, config?: AppConfig): Promise<string> {
  const { whisperUrl, removeFillerWords } = config ?? await loadRuntimeConfig();

  const form = new FormData();
  const blob = new Blob([audioData], { type: 'audio/wav' });
  form.append('file', blob, 'audio.wav');
  form.append('temperature', '0.2');
  form.append('response_format', 'json');

  if (removeFillerWords) {
    form.append(
      'prompt',
      'The following is a clear, formal transcript without any stutters, repetitions, or filler words like um and ah.'
    );
  }

  const response = await fetch(whisperUrl, {
    method: 'POST',
    body: form as unknown as BodyInit,
  });

  if (!response.ok) {
    const text = await response.text().catch(() => 'Unknown error');
    throw new Error(`Whisper API error: ${response.status} - ${text}`);
  }

  const data = await response.json();
  const rawText = typeof data === 'object' && data !== null && 'text' in data
    ? (data as { text?: unknown }).text
    : undefined;

  if (typeof rawText !== 'string') {
    console.warn('Whisper response did not include text:', data);
    return '';
  }

  let text = rawText.trim();

  if (removeFillerWords) {
    text = cleanFillerWords(text);
  }

  return text;
}

const FILLER_WORDS_PATTERN = /\b(um|uh|ah|er|hmm)\b([.,!?;])?/gi;
const DOUBLE_SPACE = /\s+/g;
const LEADING_TRAILING_SPACE = /^\s+|\s+$/g;
const PUNCTUATION_FIX = /\s+([.,!?;])/g;
const LEADING_FILLER_PUNCTUATION = /^[,\s]+/;

export function cleanFillerWords(text: string): string {
  let cleaned = text.replace(FILLER_WORDS_PATTERN, (_match, _word, punctuation: string | undefined) => {
    if (!punctuation || punctuation === ',') return '';
    return punctuation;
  });
  cleaned = cleaned.replace(DOUBLE_SPACE, ' ');
  cleaned = cleaned.replace(LEADING_TRAILING_SPACE, '');
  cleaned = cleaned.replace(PUNCTUATION_FIX, '$1');
  cleaned = cleaned.replace(LEADING_FILLER_PUNCTUATION, '');
  return cleaned;
}

async function loadRuntimeConfig(): Promise<AppConfig> {
  const { getConfig } = await import('./config');
  return getConfig();
}
